plutus-core/executables/uplc/Main.hs

```diff
@@ -204,6 +204,86 @@ runEval (EvalOptions inp ifmt printMode budgetMode traceMode outputMode timingMo
                         None -> pure ()
                         _    -> writeToFileOrStd outputMode (T.unpack (T.intercalate "\n" logs))
 
+
+-- Experimental optimisation involving force/delay.
+
+body :: UPLC.Program name uni fun ann -> UPLC.Term name uni fun ann
+body (UPLC.Program _ _ t) = t
+
+isIteratedBuiltinApp :: UPLC.Term name uni fun ann -> Bool
+isIteratedBuiltinApp =
+    \case
+      UPLC.Apply _ t _ -> isIteratedBuiltinApp t
+      UPLC.Force _ t   -> isIteratedBuiltinApp t
+      UPLC.Builtin {}  -> True
+      _                -> False
+
+
+-- Terms which can't compute, so delaying them does nothing.
+isInert :: UPLC.Term name uni fun ann -> Bool
+isInert =
+    \case
+     UPLC.Apply    {} -> False  -- Not quite true: could be a partially applied builtin.
+     UPLC.Force    {} -> False
+     UPLC.Error    {} -> False
+     UPLC.Var      {} -> False
+     UPLC.Delay    {} -> True
+     UPLC.LamAbs   {} -> True
+     UPLC.Constant {} -> True
+     UPLC.Builtin  {} -> True  -- Builtins must have at least one argument, so this can't perform any computation.
+
+-- Remove all applications of 'delay' to terms that can't compute; once we've
+-- done that, we can also retain 'force' only for terms that do compute
+-- (although for the validation examples this doesn't appear to do anything).
+-- We also have to make 'force' work on arbitrary terms in the evaluator.
+shrink :: UPLC.Term name uni fun ann -> UPLC.Term name uni fun ann
```

Comment:
you know there's a small UPLC optimizer you could have added this to? no big deal though

Reply:
> you know there's a small UPLC optimizer you could have added this to? no big deal though

I'm just going to throw it away, so I didn't bother putting everything in the right place.